src/debugger/assert.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#define ASSERT_MESGBUF_SIZE 256
+
+
+#ifndef __ASSEMBLER__
+
+extern char *__n64Assert_Filename;
+extern u32   __n64Assert_LineNum;
+extern char *__n64Assert_Condition;
+extern char __n64Assert_MessageBuf[ASSERT_MESGBUF_SIZE + 1];
+extern void __n64Assert(char *fileName, u32 lineNum, char *cond);
+
+/**
+ * Will always cause a crash with your message of choice
+ */
+#define error(message) { \
+    sprintf(__n64Assert_MessageBuf, message); \
+    __n64Assert(__FILE__, __LINE__, "Unconditional"); \
+}
+
+/**
+ * Wrapper for assert/aggress
+ */
+#define __assert_wrapper(cond) __n64Assert(__FILE__, __LINE__, (cond))
+
+/**
+ * `aggress` and `aggressf` will always cause a crash if `cond` is not true (handle with care)
+ */
+#define aggressf(cond, ...) do {\
+    if ((cond) == FALSE) { \
+        sprintf(__n64Assert_MessageBuf, __VA_ARGS__); \
+        __assert_wrapper(#cond); \
+    } \
+} while (0);
+#define aggress(cond) do {\
+    if ((cond) == FALSE) { \
+        __n64Assert_MessageBuf[0] = 0; \
+        __assert_wrapper(#cond); \
+    } \
+} while (0);
+
+/**
+ * Will cause a crash if cond is not true, and DEBUG is defined.
+ *  If disabled, `!cond` is marked as unreachable, which should
+ *  improve codegen on release builds
+ */
+#ifdef DEBUG_ASSERTIONS
+#define assertf(cond, ...) do {\
+    if ((cond) == FALSE) { \
+        sprintf(__n64Assert_MessageBuf, __VA_ARGS__); \
+        __assert_wrapper(#cond); \
+    } \
+} while (0);
+#else
+#define assertf(cond, ...) if ((cond) == FALSE) { __builtin_unreachable(); }
+#endif
+
+#ifdef DEBUG_ASSERTIONS
+#define assert(cond) do {\
+    if ((cond) == FALSE) { \
+        __n64Assert_MessageBuf[0] = 0; \
+        __assert_wrapper(#cond); \
+    } \
+} while (0);
+#else
+#define assert(cond) if ((cond) == FALSE) { __builtin_unreachable(); }
```

Comment:
Wait unless I simply don't understand the nitty gritty of GCC optimization here, why do we want this to still generate any code at all when assertions are disabled?

Reply:
Once a game has been tested to not throw any assert, the dev can disable them and the compiler can now use the condition as a hint and assume that it's always true, adding a little bit of optimization to the codebase

Reply:
That isn't necessarily good though; I use assert over aggress for non-critical failure points, but marking it unreachable would essentially be making them critical failure points.